Graduate Evented PLEG to Beta (#39913)
kubernetes/website: content/en/docs/tasks/administer-cluster/switch-to-evented-pleg.md

Reviewer: Let's not call this “evented PLEG” as this is jargon many cluster operators won't understand.

Instead, talk about switching from polling to event-based updates to container status.

Author: Updated the Title.

---
title: Switch to Evented PLEG
content_type: task
weight: 90

> ---

Reviewer: ```suggestion
min-kubernetes-server-version: 1.26
---
```

Author: Updated. Thanks.

<!-- overview -->
This page shows how to use `Evented PLEG` in Kubelet which attempts to lower the resource consumption by avoiding frequent polling for container statuses.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

> 

Author: > * Do you need to enable any feature gates on the control plane?
No.

> * Do you need a specific Kubernetes version?

1.26+

> * Do you need administrator-level access to the node(s) you're managing?

Yes

> * Do you need a container runtime that has support for a particular CRI feature?

Yes

> 
> and we should also cover, somewhere in this page:
> 
> * do you need to roll this out to all nodes at once, or can you phase it?

You can phase it. It is only a node specific feature that affects how the given kubelet manages the pod status cache internally.

<!-- steps -->

## Why switch to Evented PLEG?

* The current `Generic PLEG` incurs non-negligible overhead due to frequent polling of container statuses.
* This overhead is exacerbated by Kubelet's parallelism, limiting its scalability and causing poor performance and reliability problems.

> * The goal of `Evented PLEG` is to reduce unnecessary work during inactivity by replacing periodic polling.

Reviewer: This should go before the list of prerequisites.

> 

Author: yes.

## Switching to Evented PLEG

1. Start the `CRI Runtime` with the `Evented PLEG` support. e.g.
   1. Containerd version 1.7+ 
   2. CRI-O v1.26+ and also, start the cri-o daemon with the flag `--enable-pod-events=true` or using a drop in config like,
      ```toml
      [crio.runtime]
      enable_pod_events: true

>       ```

Author: Updated. Thanks.

---

> title: Switching from polling to CRI event-based updates to container status

Reviewer: ```suggestion
title: Switching From Polling to CRI Event-based Updates to Container Status
```

min-kubernetes-server-version: 1.26
content_type: task
weight: 90
---

{{< feature-state for_k8s_version="v1.26" state="beta" >}}

<!-- overview -->

> This page shows how to use `Evented PLEG` (CRI event-based updates to container status) in Kubelet which attempts to lower the resource consumption by avoiding frequent polling for container statuses.

Reviewer: ```suggestion
This page shows how to migrate notes to use event based updates for container status. The event-based
implementation reduces node resource consumption by the kubelet, compared to the legacy approach
that relies on polling.

You may know this feature as _evented Pod lifecycle event generator (PLEG)_. That's the name used
internally within the Kubernetes project for a key implementation detail.
```
?

## {{% heading "prerequisites" %}}

* This support for this feature is available in Kubernetes {{< skew currentVersion >}};
  Kubernetes v1.26 as a beta feature, disabled by default.

> * Please make sure the node is [drained](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) before proceeding. 

Reviewer: This should be part of the steps and not a prerequisite.

* Start the `CRI Runtime` with the `Evented PLEG` support. 
      {{< tabs name="tab_with_code" >}}
      {{< tab name="Containerd" codelang="bash" >}}
      Version 1.7+
      {{< /tab >}}
      {{< tab name="CRI-O" codelang="bash" >}}
      v1.26+ and also, start the cri-o daemon with the flag `--enable-pod-events=true` or using a drop in config like,
            
      [crio.runtime]
      enable_pod_events: true
      
      {{< /tab >}}

>       {{< /tabs >}}

Reviewer: At this point, we should have the reader check that their runtime _can_ support events. They don't have to have already reconfigured it (eg CRI-O and `enable_pod_events`).
